Libunifex does not compile on Linux

It appears that there is a namespace called `linux`. For some reason, Clang 9 #define's linux to be 1 on my platform (Ubuntu 18.04). I'm using the clang-9 from apt.llvm.org. Error below:

/usr/bin/clang++-9   -I../source/../include -Iinclude -stdlib=libc++ -std=gnu++2a -MD -MT source/CMakeFiles/unifex.dir/linux/monotonic_clock.cpp.o -MF source/CMakeFiles/unifex.dir/linux/monotonic_clock.cpp.o.d -o source/CMakeFiles/unifex.dir/linux/monotonic_clock.cpp.o -c ../source/linux/monotonic_clock.cpp
In file included from ../source/linux/monotonic_clock.cpp:16:
../source/../include/unifex/linux/monotonic_clock.hpp:24:11: error: expected identifier or '{'
namespace linux {
          ^
<built-in>:392:15: note: expanded from here
#define linux 1
              ^

--- a/cmake/FindLibUring.cmake
+++ b/cmake/FindLibUring.cmake
@@ -1,0 +1,29 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+find_path(LIBURING_INCLUDE_DIR NAMES liburing.h)
+mark_as_advanced(LIBURING_INCLUDE_DIR)
+
+find_library(LIBURING_LIBRARY NAMES uring)
+mark_as_advanced(LIBURING_LIBRARY)
+
+include(FindPackageHandleStandardArgs)
+FIND_PACKAGE_HANDLE_STANDARD_ARGS(
+        LIBURING
+        REQUIRED_VARS LIBURING_LIBRARY LIBURING_INCLUDE_DIR)
+
+if(LIBURING_FOUND)
+  set(LIBURING_LIBRARIES ${LIBURING_LIBRARY})
+  set(LIBURING_INCLUDE_DIRS ${LIBURING_INCLUDE_DIR})
+endif()

--- a/cmake/unifex_flags.cmake
+++ b/cmake/unifex_flags.cmake
@@ -29,7 +29,7 @@ else()
 endif()
 
 # Probe for libUring support
-find_package(LibUring COMPONENTS Experimental Final)
+find_package(LibUring COMPONENTS)
 # Set some variables to be used by configure_file.
 if(LIBURING_FOUND)
   set(UNIFEX_NO_LIBURING FALSE)

--- a/cmake/unifex_flags.cmake
+++ b/cmake/unifex_flags.cmake
@@ -27,3 +27,15 @@ if(CXX_MEMORY_RESOURCE_HAVE_PMR)
 else()
   set(UNIFEX_NO_MEMORY_RESOURCE TRUE)
 endif()
+
+# Probe for libUring support
+find_package(LibUring COMPONENTS Experimental Final)
+# Set some variables to be used by configure_file.
+if(LIBURING_FOUND)
+  set(UNIFEX_NO_LIBURING FALSE)
+  set(UNIFEX_URING_HEADER ${LIBURING_LIBRARIES})
+  set(UNIFEX_URING_LIBRARY ${LIBURING_INCLUDE_DIRS})
+else()
+  set(UNIFEX_NO_LIBURING TRUE)
+endif()
+

--- a/examples/linux/io_uring_test.cpp
+++ b/examples/linux/io_uring_test.cpp
@@ -13,6 +13,9 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
+#if __has_include(<liburing.h>)
+
 #include <unifex/inplace_stop_token.hpp>
 #include <unifex/just.hpp>
 #include <unifex/let.hpp>
@@ -143,3 +146,10 @@ int main() {
 
   return 0;
 }
+
+#else
+#include <cstdio>
+int main() {
+  printf("liburing support not found\n");
+}
+#endif // __has_include(<liburing.h>)

--- a/include/unifex/detail/intrusive_heap.hpp
+++ b/include/unifex/detail/intrusive_heap.hpp
@@ -32,7 +32,7 @@ class intrusive_heap {
       assert(item->*Prev == nullptr);
     }
     while (item != nullptr) {
-      std::printf("- %p\n", item);
+      printf("- %p\n", item);
       if (item->*Next != nullptr) {
         assert(item->*Next->*Prev == item);
       }

--- a/include/unifex/linux/io_uring_context.hpp
+++ b/include/unifex/linux/io_uring_context.hpp
@@ -15,6 +15,8 @@
  */
 #pragma once
 
+#if __has_include(<liburing.h>)
+
 #include <unifex/detail/atomic_intrusive_queue.hpp>
 #include <unifex/detail/intrusive_heap.hpp>
 #include <unifex/detail/intrusive_queue.hpp>
@@ -42,7 +44,7 @@
 #include <sys/uio.h>
 
 namespace unifex {
-namespace linux {
+namespace linuxos {
 
 class io_uring_context {
  public:
@@ -902,5 +904,7 @@ inline io_uring_context::scheduler io_uring_context::get_scheduler() noexcept {
   return scheduler{*this};
 }
 
-} // namespace linux
+} // namespace linuxos
 } // namespace unifex
+
+#endif // __has_include(<liburing.h>)

--- a/include/unifex/linux/mmap_region.hpp
+++ b/include/unifex/linux/mmap_region.hpp
@@ -19,7 +19,7 @@
 #include <utility>
 
 namespace unifex {
-namespace linux {
+namespace linuxos {
 
 struct mmap_region {
   mmap_region() noexcept : ptr_(nullptr), size_(0) {}
@@ -52,5 +52,5 @@ struct mmap_region {
   std::size_t size_;
 };
 
-} // namespace linux
+} // namespace linuxos
 } // namespace unifex

--- a/include/unifex/linux/monotonic_clock.hpp
+++ b/include/unifex/linux/monotonic_clock.hpp
@@ -21,7 +21,7 @@
 #include <ratio>
 
 namespace unifex {
-namespace linux {
+namespace linuxos {
 
 // A std::chrono-like clock type that wraps the Linux CLOCK_MONOTONIC clock
 // type.
@@ -190,5 +190,5 @@ inline void monotonic_clock::time_point::normalize() noexcept {
   }
 }
 
-} // namespace linux
+} // namespace linuxos
 } // namespace unifex

--- a/include/unifex/linux/safe_file_descriptor.hpp
+++ b/include/unifex/linux/safe_file_descriptor.hpp
@@ -18,7 +18,7 @@
 #include <utility>
 
 namespace unifex {
-namespace linux {
+namespace linuxos {
 
 class safe_file_descriptor {
  public:
@@ -54,5 +54,5 @@ class safe_file_descriptor {
   int fd_;
 };
 
-} // namespace linux
+} // namespace linuxos
 } // namespace unifex

--- a/include/unifex/static_thread_pool.hpp
+++ b/include/unifex/static_thread_pool.hpp
@@ -26,6 +26,8 @@
 #include <type_traits>
 #include <vector>
 #include <atomic>
+#include <mutex>
+#include <condition_variable>
 
 namespace unifex
 {

--- a/source/CMakeLists.txt
+++ b/source/CMakeLists.txt
@@ -20,13 +20,28 @@ if (CMAKE_SYSTEM_NAME STREQUAL "Linux")
     PRIVATE
       linux/mmap_region.cpp
       linux/monotonic_clock.cpp
-      linux/safe_file_descriptor.cpp
-      linux/io_uring_context.cpp)
+      linux/safe_file_descriptor.cpp)
 
   target_link_libraries(unifex
     PRIVATE
-      pthread
-      uring)
+      pthread)
+
+if (UNIFEX_NO_LIBURING EQUAL FALSE)
+
+  target_sources(unifex
+    PRIVATE
+      linux/io_uring_context.cpp)
+
+  target_include_directories(unifex
+    PUBLIC
+      ${UNIFEX_URING_HEADER})
+
+  target_link_libraries(unifex
+    PRIVATE
+      ${UNIFEX_URING_LIBRARY})
+
+endif()
+
 endif()
 
 configure_file(

--- a/source/linux/io_uring_context.cpp
+++ b/source/linux/io_uring_context.cpp
@@ -13,6 +13,9 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
+#if __has_include(<liburing.h>)
+
 #include <unifex/linux/io_uring_context.hpp>
 
 #include <unifex/scope_guard.hpp>
@@ -237,7 +240,7 @@
 // the list of remotely scheduled items and add them to the list of
 // ready-to-run operations.
 
-namespace unifex::linux {
+namespace unifex::linuxos {
 
 static thread_local io_uring_context* currentThreadContext;
 
@@ -821,4 +824,6 @@ io_uring_context::async_read_write_file tag_invoke(
   return io_uring_context::async_read_write_file{*scheduler.context_, result};
 }
 
-} // namespace unifex::linux
+} // namespace unifex::linuxos
+
+#endif // __has_include(<liburing.h>)

--- a/source/linux/mmap_region.cpp
+++ b/source/linux/mmap_region.cpp
@@ -17,7 +17,7 @@
 
 #include <sys/mman.h>
 
-namespace unifex::linux {
+namespace unifex::linuxos {
 
 mmap_region::~mmap_region() {
   if (size_ > 0) {
@@ -25,4 +25,4 @@ mmap_region::~mmap_region() {
   }
 }
 
-} // namespace unifex::linux
+} // namespace unifex::linuxos

--- a/source/linux/monotonic_clock.cpp
+++ b/source/linux/monotonic_clock.cpp
@@ -17,7 +17,7 @@
 
 #include <time.h>
 
-namespace unifex::linux {
+namespace unifex::linuxos {
 
 monotonic_clock::time_point monotonic_clock::now() noexcept {
   timespec ts;
@@ -25,4 +25,4 @@ monotonic_clock::time_point monotonic_clock::now() noexcept {
   return time_point::from_seconds_and_nanoseconds(ts.tv_sec, ts.tv_nsec);
 }
 
-} // namespace unifex::linux
+} // namespace unifex::linuxos

--- a/source/linux/safe_file_descriptor.cpp
+++ b/source/linux/safe_file_descriptor.cpp
@@ -19,7 +19,7 @@
 
 #include <unistd.h>
 
-namespace unifex::linux {
+namespace unifex::linuxos {
 
 void safe_file_descriptor::close() noexcept {
   assert(valid());
@@ -27,4 +27,4 @@ void safe_file_descriptor::close() noexcept {
   assert(result == 0);
 }
 
-} // namespace unifex::linux
+} // namespace unifex::linuxos

